src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1821,4 +1821,97 @@ private final void appendChars(CharSequence s, int off, int end) {
         }
         count += end - off;
     }
+
+    /**
+     * Appends {@code count} copies of the character {@code c} to this sequence.
+     * <p>
+     * The length of this sequence increases by {@code count}.
+     *
+     * @param c      character to append
+     * @param count  number of times to copy
+     *
+     * @return  a reference to this object.
+     *
+     * @since 21
+     * @throws IllegalArgumentException  if {@code count} is less than zero
+     * @throws IndexOutOfBoundsException  if the result overflows the buffer
+     */
+    public AbstractStringBuilder repeat(char c, int count) {
+        if (count < 0) {
+            throw new IllegalArgumentException("count is less than zero: " + count);
+        } else if (count == 0) {
+            return this;
+        }
+        ensureCapacityInternal(this.count + count);
+        int index = this.count;
+        int limit = index + count;
+        boolean isLatin1 = isLatin1();
+        if (isLatin1 && StringLatin1.canEncode(c)) {
+            while (index < limit) {
+                value[index++] = (byte)c;
+            }
+        } else {
+            if (isLatin1) {
+                inflate();
+            }
+            while (index < limit) {
+                StringUTF16.putCharSB(value, index++, c);
+            }
+        }
+        this.count = limit;
+        return this;
+    }
+
+    /**
+     * Appends {@code count} copies of the specified {@code CharSequence} {@code cs}
+     * to this sequence.
+     * <p>
+     * The length of this sequence increases by {@code count} times the
+     * {@code CharSequence} length.
+     *
+     * @param cs     a {@code CharSequence}
+     * @param count  number of times to copy
+     *
+     * @return  a reference to this object.
+     *
+     * @since 21
+     * @throws IllegalArgumentException  if {@code count} is less than zero
+     * @throws IndexOutOfBoundsException  if the result overflows the buffer
+     */
+    public AbstractStringBuilder repeat(CharSequence cs, int count) {
+        if (count < 0) {
+            throw new IllegalArgumentException("count is less than zero: " + count);
+        } else if (count == 0) {
+            return this;
+        } else if (count == 1) {
+            return append(cs);
+        }
+        if (cs == null) {
+            cs = "null";
+        }
+        int length = cs.length();
+        if (length == 0) {
+            return this;
+        } else if (length == 1) {
+            return repeat(cs.charAt(0), count);
+        }
+        int offset = this.count;
+        int valueLength = length << coder;
+        if ((Integer.MAX_VALUE - offset) / count < valueLength) {
+            throw new OutOfMemoryError("Required length exceeds implementation limit");
+        }
+        int total = count * length;
```

Comment:
We may avoid division if we use the long type:

```java
long totalLong = ((long) count) * length;
if (totalLong > Integer.MAX_VALUE - offset) {
    throw new OutOfMemoryError("Required length exceeds implementation limit");
}
int total = (int) totalLong;
```

Should be faster.

Reply:
I'm a bit surprised this (and the original code) is throwing `OutOfMemoryError` when running into the max array size. It is not truly an out of memory error, but being an `Error`, it would evade standard catch `Exception` blocks.  I would think this is more of an `IllegalStateException` or perhaps something array specific.

`OutOfMemoryError` is documented pretty specifically that an object allocation failed after exhaustive GC, but no allocation or GC happened:

>Thrown when the Java Virtual Machine cannot allocate an object because it is out of memory, and no more memory could be made available by the garbage collector.

... which is not happening here at all. This leads me to believe the error is not used correctly here.

Other reasons I find it surprising: `StringBuilder` is not documented to throw this anywhere, it seems to just leak through from various methods implemented by `AbstractStringBuilder`.

Reply:
OOME is used for a number of not-strictly out of memory conditions, for example off-heap allocation and allocations that can't succeed because they exceed implementation limits.